File: yurt-system/yurt-system-server/src/main/java/com/github/liulus/yurt/system/service/impl/ParamServiceImpl.java
package com.github.liulus.yurt.system.service.impl;

import com.github.liulus.yurt.convention.data.Page;
import com.github.liulus.yurt.convention.exception.ServiceException;
import com.github.liulus.yurt.system.model.dto.SysParamQo;
import com.github.liulus.yurt.system.model.entity.SysParam;
import com.github.liulus.yurt.system.repository.ParamRepository;
import com.github.liulus.yurt.system.service.ParamService;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import javax.annotation.Resource;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

/**
 * User : liulu
 * Date : 17-9-17 下午2:53
 * version $Id: ParamServiceImpl.java, v 0.1 Exp $
 */
@Service
@Transactional
public class ParamServiceImpl implements ParamService {

    @Resource
    private ParamRepository paramRepository;

    @Override
    public Page<SysParam> findPageList(SysParamQo qo) {
//        return paramRepository.selectPageList(qo);
        return null;
    }

    @Override
    public SysParam findById(Long id) {
        return paramRepository.selectById(id);
    }

    @Override
    public SysParam findByCode(String code) {
        return paramRepository.selectByCode(code);
    }

    @Override
    public Long insert(SysParam param) {
        checkCode(param.getCode());
        param.setSystem(false);

        paramRepository.insert(param);
        return param.getId();
    }

    @Override
    public void update(SysParam param) {
        SysParam oldParam = findById(param.getId());
        if (oldParam == null) {
            return;
        }
        if (!Objects.equals(param.getCode(), oldParam.getCode())) {
            if (oldParam.getSystem()) {
                throw new ServiceException(String.format("%s 是系统级参数, 不允许修改", oldParam.getCode()));
            }
            checkCode(param.getCode());
        }
        param.setSystem(null);
        paramRepository.updateIgnoreNull(param);
    }

    private void checkCode(String code) {
        SysParam param = findByCode(code);
        if (param != null) {
            throw new ServiceException("参数 code 已经存在");
        }
    }

    @Override
    public void delete(Long... ids) {
        if (ids == null || ids.length == 0) {
            return;
        }

        List<SysParam> params = paramRepository.selectByIds(Arrays.asList(ids));

        List<Long> validIds = new ArrayList<>(params.size());

        for (SysParam param : params) {
            if (param.getSystem()) {
                throw new ServiceException(String.format("%s 是系统级参数, 不允许删除", param.getCode()));
            }
            validIds.add(param.getId());
        }
//        paramRepository.deleteByIds(validIds);
    }
}

File: yurt-system/yurt-system-server/src/main/kotlin/com/github/liulus/yurt/system/service/impl/ParamServiceImpl.kt
package com.github.liulus.yurt.system.service.impl

import com.github.liulus.yurt.convention.data.Page
import com.github.liulus.yurt.convention.exception.ServiceException
import com.github.liulus.yurt.system.model.dto.SysParamQo
import com.github.liulus.yurt.system.model.entity.SysParam
import com.github.liulus.yurt.system.repository.ParamRepository
import com.github.liulus.yurt.system.service.ParamService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import javax.annotation.Resource

/**
 * User : liulu
 * Date : 17-9-17 下午2:53
 * version $Id: ParamServiceImpl.java, v 0.1 Exp $
 */
@Service
@Transactional
open class ParamServiceImpl : ParamService {
    @Resource
    private val paramRepository: ParamRepository? = null
    override fun findPageList(qo: SysParamQo?): Page<SysParam?>? {
//        return paramRepository.selectPageList(qo);
        return null
    }

    override fun findById(id: Long?): SysParam? {
        return paramRepository!!.selectById(id!!)
    }

    override fun findByCode(code: String?): SysParam? {
        return paramRepository!!.selectByCode(code)
    }

    override fun insert(param: SysParam?): Long? {
        checkCode(param!!.code)
        param.system = false
        paramRepository!!.insert(param)
        return param.id
    }

    override fun update(param: SysParam?) {
        val (_, code, _, _, system) = findById(param!!.id) ?: return
        if (param.code != code) {
            if (system!!) {
                throw ServiceException(String.format("%s 是系统级参数, 不允许修改", code))
            }
            checkCode(param.code)
        }
        param.system = null
        paramRepository!!.updateIgnoreNull(param)
    }

    private fun checkCode(code: String?) {
        val param = findByCode(code)
        if (param != null) {
            throw ServiceException("参数 code 已经存在")
        }
    }

//    override fun delete(vararg ids: Long) {
//        if (ids == null || ids.size == 0) {
//            return
//        }
//        val params = paramRepository!!.selectByIds(Arrays.asList(*ids))
//        val validIds: MutableList<Long?> = ArrayList(params.size)
//        for ((id, code, _, _, system) in params) {
//            if (system!!) {
//                throw ServiceException(String.format("%s 是系统级参数, 不允许删除", code))
//            }
//            validIds.add(id)
//        }
//        //        paramRepository.deleteByIds(validIds);
//    }
}